contracts/moda-contracts/src/ModaRegistry.sol

```diff
@@ -3,22 +3,158 @@ pragma solidity 0.8.21;
 
 import "./interfaces/IModaRegistry.sol";
 import "./interfaces/IMembership.sol";
+import "@openzeppelin/contracts/access/extensions/AccessControlEnumerable.sol";
 
-contract ModaRegistry {
-    mapping(string => address) _catalogMemberships;
+contract ModaRegistry is IModaRegistry, AccessControlEnumerable {
+    using EnumerableSet for EnumerableSet.AddressSet;
 
+    /// State Variables
 
-    function isMember(string calldata catalogName, address user) external view returns (bool) {
-        address membership = _catalogMemberships[catalogName];
+    uint256 _catalogCount;
+    address payable _treasury;
+    uint256 _treasuryFee;
 
-        if (IMembership(membership).isMember(user)) {
+    mapping(string => Catalog) _catalogs;
+    mapping(address => EnumerableSet.AddressSet) _managers;
+
+    /// Errors
+
+    error CatalogAlreadyRegistered();
+    error CatalogDoesNotExist();
+    error NoCurrentCatalogs();
+
+    constructor() {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    /// Membership
+
+    /**
+     * @inheritdoc IModaRegistry
+     */
+    function isMember(string calldata name, address user) external view returns (bool) {
+        if (IMembership(_catalogs[name].membership).isMember(user)) {
             return true;
         } else {
             return false;
         }
     }
 
-    function registerCatalog(string calldata catalog, address catalogAddress) external {
-        _catalogMemberships[catalog] = catalogAddress;
+    /**
+     * @inheritdoc IModaRegistry
+     */
+    function setCatalogMembership(string calldata name, address membership) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _catalogs[name].membership = membership;
+        emit CatalogMembershipSet(name, membership);
+    }
+
+    /**
+     * @inheritdoc IModaRegistry
+     */
+    function registerCatalog(string calldata name, address catalog) external {
+        if (_catalogs[name].registered) {
+            revert CatalogAlreadyRegistered();
+        }
+        _catalogs[name].catalog = catalog;
+        _catalogs[name].registered = true;
+        _catalogCount++;
+
+        emit CatalogRegistered(name, catalog);
+    }
+
+    /**
+     * @inheritdoc IModaRegistry
+     */
+    function unregisterCatalog(string calldata name) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (_catalogs[name].catalog == address(0)) {
+            revert CatalogDoesNotExist();
+        }
+        if (_catalogCount == 0) {
+            revert NoCurrentCatalogs();
+        }
+        _catalogs[name].registered = false;
+        _catalogCount--;
+        emit CatalogUnregistered(name, _catalogs[name].catalog);
+    }
+
+    /**
+     * @inheritdoc IModaRegistry
+     */
+    function getCatalog(string calldata name) external view returns (Catalog memory) {
```

Comment:
It could be dangerous if we unregister a malicious catalog then people use this method without checking `isRegistered`. It probably makes sense to set zero values to everything when we unregister